GAE/src/org/waterforpeople/mapping/dao/SurveyInstanceDAO.java

```diff
@@ -139,13 +139,8 @@ public SurveyInstance save(SurveyInstance si, DeviceFiles deviceFile) {
             String filename = qas.getValue().substring(
                     qas.getValue().lastIndexOf("/") + 1);
 
-            Device d = null;
-            if (deviceFile.getImei() != null) {
-                d = deviceDao.getByImei(deviceFile.getImei());
-            }
-            if (d == null && deviceFile.getPhoneNumber() != null) {
-                d = deviceDao.get(deviceFile.getPhoneNumber());
-            }
+            Device d = deviceDao.getDevice(deviceFile.getAndroidId(), 
```

Comment:
I noticed that this is within a loop. Perhaps you can shift it out and instead use a hash to store the device (if thats necessary at all).